agents/common/nexagent.go

```diff
@@ -1,34 +1,169 @@
 package agentcommon
 
-import "github.com/nats-io/nats.go"
+import (
+	"encoding/json"
+	"fmt"
+	"log/slog"
+	"os"
+	"time"
+
+	"github.com/nats-io/nats.go"
+	agentapi "github.com/synadia-io/nex/api/agent/go"
+	agentapigen "github.com/synadia-io/nex/api/agent/go/gen"
+)
 
 // A convenient bundling for common functions that belong to a Nex Agent
 type NexAgent struct {
-	upHandler        CommandHandler
-	preflightHandler CommandHandler
-	embeddedNats     *nats.Conn
+	embeddedNats *nats.Conn
+	callback     AgentCallback
+
+	name         string
+	version      string
+	description  string
+	maxWorkloads int
+
+	logger *slog.Logger
 }
 
-func NewNexAgent(up CommandHandler, preflight CommandHandler) (*NexAgent, error) {
+// TODO: these callbacks need appropriate parameters
+type AgentCallback interface {
+	Up() error
+	Preflight() error
+	StartWorkload(request *agentapigen.StartWorkloadRequestJson) error
+	StopWorkload(request *agentapigen.StopWorkloadRequestJson) error
+	ListWorkloads() error
+}
+
+func NewNexAgent(name, version, description string,
+	maxWorkloads int, logger *slog.Logger, callback AgentCallback) (*NexAgent, error) {
+
 	// TODO: generate whatever context we need
 
 	embedded, err := CreateEmbeddedNatsConnection()
 	if err != nil {
 		return nil, err
 	}
-	return &NexAgent{upHandler: up, preflightHandler: preflight, embeddedNats: embedded}, nil
+	return &NexAgent{
+		logger:       logger,
+		embeddedNats: embedded,
+		callback:     callback,
+		name:         name,
+		version:      version,
+		description:  description,
+		maxWorkloads: maxWorkloads,
+	}, nil
 }
 
 func (agent *NexAgent) Run() error {
 	// TODO: based on argv, decide which handler to call
 	// get error from handler and return here
+
+	if err := agent.registerWithNode(); err != nil {
+		return err
+	}
+
+	if err := agent.createSubscriptions(); err != nil {
+		return err
+	}
+
+	switch os.Args[1] {
```

Comment:
What if we dont pass it a bunch of args here and instead we pass it a json string with all the parameters it's expecting?  Seems easier than parsing the command line

Reply:
there's only 1 command line parameter here - `up` or `preflight`. The reason they're command line switches is so that they can be executed in a variety of ways, including by the `nex node preflight` command, which will eventually locate all the agents and run them

Reply:
every other parameter the agent needs comes from an environment variable